src/parser/jobs/dnc/modules/DirtyDancing.tsx

```diff
@@ -0,0 +1,298 @@
+	ACTIONS.DOUBLE_TECHNICAL_FINISH.id,
+	ACTIONS.TRIPLE_TECHNICAL_FINISH.id,
+	ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id,
+]
+
+const FINISHES = [
+	...STANDARD_FINISHES,
+	...TECHNICAL_FINISHES,
+]
+
+const EXPECTED_DANCE_MOVE_COUNT = {
+	[ACTIONS.DOUBLE_STANDARD_FINISH.id]: 2,
+	[ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id]: 4,
+}
+
+class Dance {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+	dancing: boolean = false
+	resolved: boolean = false
+
+	dirty: boolean = false
+	missed: boolean = false
+	footloose: boolean = false
+
+	public get error(): boolean {
+		return this.dirty || this.missed || this.footloose
+	}
+
+	public get expectedFinishId(): number {
+		const actualFinish = _.last(this.rotation)
+		let expectedFinish = -1
+		if (actualFinish) {
+			if (TECHNICAL_FINISHES.includes(actualFinish.ability.guid)) {
+				expectedFinish =  ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id
+			} else if (STANDARD_FINISHES.includes(actualFinish.ability.guid)) {
+				expectedFinish = ACTIONS.DOUBLE_STANDARD_FINISH.id
+			}
+		}
+		return expectedFinish
+	}
+
+	constructor(start: number) {
+		this.start = start
+		this.dancing = true
+	}
+}
+
+export default class DirtyDancing extends Module {
+	static handle = 'dirtydancing'
+	static title = t('dnc.dirty-dancing.title')`Dance Issues`
+
+	@dependency private checklist!: CheckList
+	@dependency private suggestions!: Suggestions
+	@dependency private invuln!: Invulnerability
+	@dependency private combatants!: Combatants
+	@dependency private timeline!: Timeline
+
+	private danceHistory: Dance[] = []
+	private missedDances = 0
+	private dirtyDances = 0
+	private footlooseDances = 0
+
+	protected init() {
+		this.addHook('cast', {by: 'player', abilityId: STEPS}, this.beginDance)
+		this.addHook('cast', {by: 'player'}, this.continueDance)
+		this.addHook('cast', {by: 'player', abilityId: FINISHES}, this.finishDance)
+		this.addHook('damage', {by: 'player', abilityId: FINISHES}, this.resolveDance)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private addDanceToHistory(event: CastEvent): Dance {
+		const newDance = new Dance(event.timestamp)
+		newDance.rotation.push(event)
+		this.danceHistory.push(newDance)
+
+		return newDance
+	}
+
+	private beginDance(event: CastEvent) {
+		this.addDanceToHistory(event)
+	}
+
+	private get lastDance(): Dance | undefined {
+		return _.last(this.danceHistory)
+	}
+
+	private continueDance(event: CastEvent) {
+		if (!STEPS.includes(event.ability.guid) && !FINISHES.includes(event.ability.guid)) {
+			const dance = this.lastDance
+			if (dance && dance.dancing) {
+				dance.rotation.push(event)
+			}
+		}
+	}
+
+	private finishDance(event: CastEvent) {
+		let dance = this.lastDance
+		if (dance && dance.dancing) {
+			dance.rotation.push(event)
+		} else {
+			dance = this.addDanceToHistory(event)
+		}
+		dance.dancing = false
+	}
+
+	private resolveDance(event: DamageEvent) {
+		const dance = this.lastDance
+		if (dance && !dance.resolved) {
```

Comment:
likewise; can shortcut out here so you don't need to lock the rest inside the block